Client/TabManagement/TabManagerImplementation.swift

```diff
@@ -3,7 +3,117 @@
 // file, You can obtain one at http://mozilla.org/MPL/2.0/
 
 import Foundation
+import TabDataStore
+import Storage
+import Common
+
+// Hard coding a feature flag while the feature is in development
+// This will be replaced with a nimbus flag when the feature is ready
+// FXIOS-5885
+var useNewTabDataStore = false
 
 // This class subclasses the legacy tab manager temporarily so we can
 // gradually migrate to the new system
-class TabManagerImplementation: LegacyTabManager {}
+class TabManagerImplementation: LegacyTabManager {
+    let tabDataStore: TabDataStore
+
+    init(profile: Profile,
+         imageStore: DiskImageStore?,
+         logger: Logger = DefaultLogger.shared,
+         tabDataStore: TabDataStore = DefaultTabDataStore()) {
+        self.tabDataStore = tabDataStore
+        super.init(profile: profile, imageStore: imageStore)
+    }
+
+    // MARK: - Restore tabs
+
+    override func restoreTabs(_ forced: Bool = false) {
+        guard useNewTabDataStore else {
+            super.restoreTabs(forced)
+            return
+        }
+
+        guard !isRestoringTabs else { return }
+
+        // TODO: FXIOS-6112 Handle debug settings and UITests
+
+        if forced {
+            tabs = [Tab]()
+        }
+
+        isRestoringTabs = true
+        Task {
+            guard let windowData = await self.tabDataStore.fetchTabData() else {
+                // Always make sure there is a single normal tab
+                self.generateEmptyTab()
+                return
+            }
+
+            await self.generateTabs(from: windowData)
+
+            for delegate in self.delegates {
+                delegate.get()?.tabManagerDidRestoreTabs(self)
+            }
+
+            self.isRestoringTabs = false
+        }
+    }
+
+    @MainActor
+    private func generateTabs(from windowData: WindowData) async {
+        for tabData in windowData.tabData {
+            let newTab = addTab(flushToDisk: false, zombie: true, isPrivate: tabData.isPrivate)
+            newTab.url = URL(string: tabData.siteUrl)
+            newTab.lastTitle = tabData.title
+            newTab.tabUUID = tabData.id.uuidString
+            newTab.screenshotUUID = tabData.id
+            newTab.firstCreatedTime = tabData.createdAtTime.toTimestamp()
+            newTab.sessionData = LegacySessionData(currentPage: 0,
```

Comment:
That's using LegacySessionData and LegacyTabGroupData, so this will use the new classes soon? Or that's meant to stay as is?

Reply:
The Tab class will remain as it is for phase 1 of the project. This separates concerns so the data storage layer is no longer aware of the Tab class, which contains view logic and all sorts of stuff the data layer shouldn't have access to.